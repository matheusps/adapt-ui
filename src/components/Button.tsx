import React, { FC } from 'react'
import styled from 'styled-components'

import { getMeasure } from '../helpers'
import { Loader } from './Loader'
import { Clickable } from './core/Clickable'

const getFontSize = (measure: Measure): number =>
  getMeasure(measure, [0.8, 1, 1.25, 1.5])

const getPadding = (measure: Measure): string =>
  getMeasure(measure, [
    '0.375rem 1.4rem 0.375rem 1.4rem',
    '0.45rem 1.8rem 0.45rem 1.8rem',
    '0.65rem 2.2rem 0.65rem 2.2rem',
    '0.80rem 2.6rem 0.80rem 2.6rem',
  ])

const StyledClickable = styled(Clickable)<any>(({ fontSize, padding }) => ({
  fontSize,
  padding,
}))

const Button: FC<ButtonType> = ({ size, children, loading, ...props }) => {
  const fontSize = `${getFontSize(size!)}rem`
  const padding = getPadding(size!)

  return (
    <StyledClickable fontSize={fontSize} padding={padding} {...props}>
      {loading ? <Loader size={getFontSize(size!)} /> : children}
    </StyledClickable>
  )
}

Button.defaultProps = {
  size: 'md',
  loading: false,
}

export { Button }
