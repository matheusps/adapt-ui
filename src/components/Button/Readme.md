Colors

```js
  <Button size="sm" shade="lighten">
    Button SM
  </Button>
  <Button size="md" shade="lighten">
    Button MD
  </Button>
  <Button size="lg">Button LG</Button>
  <Button size="xl" shade="darken">
    Button XL
  </Button>
```
