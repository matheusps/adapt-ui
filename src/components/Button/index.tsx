import { get } from 'lodash'
import styled from 'styled-components'

import {
  getMeasure,
  lightenOrDarken,
  isBrightColor,
} from '../../global/helpers'

interface Props extends withGoal {
  size?: Measure
  kind?: 'link' | 'ghost' | 'default'
}

const getFontSize = (measure: Measure) =>
  getMeasure(measure, ['0.8rem', '1rem', '1.25rem', '1.50rem'])

const getPadding = (measure: Measure) =>
  getMeasure(measure, [
    '0.375rem 1.4rem 0.375rem 1.4rem',
    '0.45rem 1.8rem 0.45rem 1.8rem',
    '0.65rem 2.2rem 0.65rem 2.2rem',
    '0.80rem 2.6rem 0.80rem 2.6rem',
  ])

const Button = styled.button<Props>`
  background-color: ${({ kind, theme, goal }) =>
    kind === 'ghost' || kind === 'link'
      ? 'transparent'
      : get(theme.colors.goal, goal)};
  color: ${({ kind, theme, goal }) =>
    kind === 'ghost' || kind === 'link'
      ? get(theme.colors.goal, goal)
      : isBrightColor(get(theme.colors.goal, goal))
      ? '#000'
      : '#fff'};
  font-size: ${props => getFontSize(props.size!)};
  padding: ${props => getPadding(props.size!)};
  border-radius: ${({ theme }) => theme.elements.roundness};
  border: none;

  box-sizing: border-box;

  font-weight: 500;
  font-stretch: normal;
  line-height: 1.4;
  margin: 0.2rem;
  display: relative;
  overflow: hidden;
  cursor: pointer;
  text-align: center;
  text-decoration: none;
  text-transform: none;
  white-space: nowrap;
  justify-content: center;
  align-items: center;

  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;

  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;

  transition: background-color 0.2s ease-in-out;
  will-change: background-color;

  :hover {
    background-color: ${props =>
      lightenOrDarken(get(props.theme.colors.goal, props.goal), -20)};
  }

  :active {
    background-color: ${props =>
      lightenOrDarken(get(props.theme.colors.goal, props.goal), -40)};
  }

  :focus {
    outline: none;
  }
`

Button.defaultProps = {
  goal: 'create',
  size: 'md',
  kind: 'default',
}

export default Button
