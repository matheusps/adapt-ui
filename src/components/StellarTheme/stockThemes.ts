export const light = {
  colors: {
    base: {
      default: '#fafafa',
      darken: '#cecece',
      lighten: '#ffffff',
      contrast: '#000000',
    },
    empty: {
      default: '#',
      darken: '#',
      lighten: '#',
      contrast: '#',
    },
    primary: {
      default: '#256EFF',
      darken: '#113274',
      lighten: '#AFCAFF',
      contrast: '#',
    },
    secondary: {
      default: '#46237A',
      darken: '#',
      lighten: '#',
      contrast: '',
    },
    success: {
      default: '#3DDC97',
      darken: '#',
      lighten: '#',
      contrast: '',
    },
    warning: {
      default: '#',
      darken: '#',
      lighten: '#',
      contrast: '#',
    },
    info: {
      default: '#',
      darken: '#',
      lighten: '#',
      contrast: '#',
    },
    error: {
      default: '#FF495C',
      darken: '#',
      lighten: '#',
      contrast: '#',
    },
  },
}
