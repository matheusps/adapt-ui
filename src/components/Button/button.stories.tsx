import React from 'react'
import Button from './index'
import { AdaptProvider, Box } from '..'

export default {
  title: 'Button',
  decorators: [storyFn => <AdaptProvider>{storyFn()}</AdaptProvider>],
}

export const size = () => (
  <Box>
    <Button size="sm">Small</Button>
    <Button size="md">Medium</Button>
    <Button size="lg">Large</Button>
    <Button size="xl">XLarge</Button>
  </Box>
)

export const loading = () => (
  <Box>
    <Button size="sm" loading>
      Primary
    </Button>
    <Button size="md" loading>
      Secondary
    </Button>
    <Button size="lg" loading>
      Success
    </Button>
    <Button size="xl" loading>
      Warning
    </Button>
    <Button size="xl" loading>
      Danger
    </Button>
  </Box>
)

export const full = () => (
  <Box>
    <Button size="sm" full>
      Small
    </Button>
    <Button size="md" full>
      Medium
    </Button>
    <Button size="lg" full>
      Large
    </Button>
    <Button size="xl" full>
      XLarge
    </Button>
    <Button size="xl" loading full>
      XLarge
    </Button>
  </Box>
)

export const disabled = () => (
  <Box>
    <Button size="sm" disabled>
      Small
    </Button>
    <Button size="md" disabled>
      Medium
    </Button>
    <Button size="lg" disabled>
      Large
    </Button>
    <Button size="xl" disabled loading>
      XLarge
    </Button>
  </Box>
)
