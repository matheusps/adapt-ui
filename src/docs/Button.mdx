---
## name: Button
---

import { Playground } from 'docz'
import Button from '../components/Button'
import Surface from '../components/Surface'

# Button

## Size

<Playground>
  <Surface lifting={3}>
    <Button size="sm">Small</Button>
    <Button size="md">Medium</Button>
    <Button size="lg">Large</Button>
    <Button size="xl">XLarge</Button>
  </Surface>
</Playground>

## Goals

<Playground>
  <Button goal="create">Create</Button>
  <Button goal="destroy">Destroy</Button>
  <Button goal="info">Info</Button>
  <Button goal="warn">Warn</Button>
</Playground>

## Kinds

### Default

<Playground>
  <Button kind="default" goal="create">
    Create
  </Button>
  <Button kind="default" goal="destroy">
    Destroy
  </Button>
  <Button kind="default" goal="info">
    Info
  </Button>
  <Button kind="default" goal="warn">
    Warn
  </Button>
</Playground>

### Ghost

<Playground>
  <Button kind="ghost" goal="create">
    Create
  </Button>
  <Button kind="ghost" goal="destroy">
    Destroy
  </Button>
  <Button kind="ghost" goal="info">
    Info
  </Button>
  <Button kind="ghost" goal="warn">
    Warn
  </Button>
</Playground>
