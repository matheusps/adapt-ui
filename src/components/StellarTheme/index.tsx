import React, { FunctionComponent } from 'react'
import { ThemeProvider } from 'styled-components'

import { light } from './stockThemes'

/**
 * Wraps around Theme-provider from styled-components 💅🏻
 */
const StellarTheme: FunctionComponent<StellarThemeProps> = ({ children }) => {
  return (
    <ThemeProvider theme={light}>
      <>{children}</>
    </ThemeProvider>
  )
}

export default StellarTheme
