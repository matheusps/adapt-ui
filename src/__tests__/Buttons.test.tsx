import React from 'react'
import { render } from 'react-testing-library'
import Button from '../components/Button'
import StellarTheme from '../components/StellarTheme'

describe('Container component', () => {
  const renderComponent = (customProps?) =>
    render(
      <StellarTheme>
        <Button {...customProps!}>Test</Button>
      </StellarTheme>
    )

  it('should render', () => {
    const component = renderComponent()
    expect(component).toBeTruthy()
  })

  it('should match snapshot as default', () => {
    const component = renderComponent()
    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should match snapshot with all props', () => {
    const component = renderComponent({
      size: 'lg',
      color: 'primary',
    })
    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should match snapshot with all measures', () => {
    const sm = renderComponent({
      size: 'sm',
    })
    const md = renderComponent({
      size: 'md',
    })
    const lg = renderComponent({
      size: 'lg',
    })
    const xl = renderComponent({
      size: 'xl',
    })
    expect(sm.asFragment()).toMatchSnapshot()
    expect(md.asFragment()).toMatchSnapshot()
    expect(lg.asFragment()).toMatchSnapshot()
    expect(xl.asFragment()).toMatchSnapshot()
  })
})
