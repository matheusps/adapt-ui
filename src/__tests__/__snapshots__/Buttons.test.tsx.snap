// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container component should match snapshot as default 1`] = `
<DocumentFragment>
  <button
    class="sc-bdVaJa duWJrr"
    color="primary"
  >
    Test
  </button>
</DocumentFragment>
`;

exports[`Container component should match snapshot with all measures 1`] = `
<DocumentFragment>
  <button
    class="sc-bdVaJa cPnXVE"
    color="primary"
  >
    Test
  </button>
</DocumentFragment>
`;

exports[`Container component should match snapshot with all measures 2`] = `
<DocumentFragment>
  <button
    class="sc-bdVaJa duWJrr"
    color="primary"
  >
    Test
  </button>
</DocumentFragment>
`;

exports[`Container component should match snapshot with all measures 3`] = `
<DocumentFragment>
  <button
    class="sc-bdVaJa feqUcQ"
    color="primary"
  >
    Test
  </button>
</DocumentFragment>
`;

exports[`Container component should match snapshot with all measures 4`] = `
<DocumentFragment>
  <button
    class="sc-bdVaJa eCwtkd"
    color="primary"
  >
    Test
  </button>
</DocumentFragment>
`;

exports[`Container component should match snapshot with all props 1`] = `
<DocumentFragment>
  <button
    class="sc-bdVaJa feqUcQ"
    color="primary"
  >
    Test
  </button>
</DocumentFragment>
`;
